Add LLM workflow planner / builder and field mapping

muthukumaranR opened the pull request:
## Overview

This PR adds a LLM-based workflow planner that converts natural language research requests into executable workflows (compatible with AKD Backend) with automatic data routing between agents.

## Problem

When workflows need multiple agent interactions (which are composed by LLM) - and when chaining agents in a workflow (e.g., deep_search → gap_analysis), we need to:
1. Convert LLM plans into executable workflow JSON
2. Route data from one agent's outputs to the next agent's inputs
3. Handle field name mismatches (e.g., `results` → `search_results`)
4. Support both manual mappings and LLM-generated mappings

## Solution

### LLM Workflow Planner

Interactive conversational planner that:
- Understands research requirements through dialogue
- Selects appropriate agents from registry
- Extracts agent inputs using **full conversation context**:
  - Initial user request
  - Conversation history and refinements
  - Research goal and workflow plan
  - Agent selection reasoning
  - Field schemas and dependencies

**7-Phase Conversation Flow**:
1. Initial Requirements - Understand research goal
2. Goal Clarification - Refine objectives
3. Agent Selection - Choose appropriate agents
4. IO Specification - Define inputs/outputs
5. Workflow Construction - Build execution plan
6. Validation - Verify completeness
7. Finalization - Generate executable workflow

### Workflow Builder

Transforms high-level plans into executable workflows:
- Reads agent schemas from the registry (PR #236)
- Generates JSONPath expressions for runtime data access
- Creates complete workflow JSON ready for execution

### Field Mapping System

**Three-tier strategy for mapping fields**:
1. **Explicit**: Human-defined in field_mappings.json
2. **Exact Match**: Auto-match when names are identical
3. **LLM Semantic**: AI-generated with confidence scoring

### Why JSONPath?

**Problem**: At runtime, how does gap_analysis get deep_search's results?

**Without JSONPath** (hardcoded):
```python
# Custom logic for every agent pair
gap_analysis_input = {
    "search_results": deep_search_output["results"]
}
```

**With JSONPath** (declarative):
```json
{
  "type": "gap_analysis",
  "io_map": {
    "search_results": "$.deep_search.outputs.results"
  }
}
```

The executor reads the JSONPath and fetches data from the runtime state - no custom code needed per agent pair.

## Implementation

### Core Components

**LLM Planner** (NEW):
- `llm_planner.py`: Conversational workflow planning
- `LLMWorkflowPlanner`: Main planner class using LiteLLMInstructorBaseAgent (multi-provider support)
- `InteractivePlannerSession`: Manages conversation state
- Input extraction using full conversation context
- Automatic message trimming to prevent token limit errors

**Workflow Builder**:
- `format_builder.py`: Data models (WorkflowFormat, WorkflowNode, etc.)
- `workflow_builder.py`: Workflow generation with auto-generated io_map
- `structures.py`: Shared data structures (WorkflowPlan, AgentSuggestion)

**Field Mapping**:
- `field_mapping_registry.py`: Mapping storage and resolution
- `field_mapping_generator.py`: LLM-based semantic mapping

**Documentation** (NEW):
- `akd/planner/README.md`: Comprehensive documentation
- Architecture overview with diagrams
- End-to-end workflow examples
- Configuration guide

**Demo** (NEW):
- `scripts/demo_planner.py`: Interactive CLI demo
- Modes: interactive, quick, agents, sample

### Validation System

**WorkflowFormat Validation** (format_builder.py):
- `validate_io_map()`: Validates JSONPath references (format, node existence)
- `validate_edges()`: Validates graph structure (START/END, orphaned nodes)
- `validate()`: Validation with categorized results
- `get_node_summary()`: Workflow statistics (nodes, edges, io_map entries)

**Conversation Control** (llm_planner.py):
- Context-aware `ready_to_generate` validator prevents infinite loops
- Auto-corrects when contradiction exists (plan ready, no question, flag false)
- Preserves LLM decision-making for confirmation requests and clarifications

**Enum Normalization**:
- Field validators normalize enum values to lowercase
- Handles LLM uppercase outputs

### io_map Generation Example

```python
# Input: field mapping metadata
"deep_search->gap_analysis": {
  "search_results": "results"
}

# Output: JSONPath for runtime
io_map = {
  "search_results": "$.deep_search.outputs.results"
}
```

## Tests

**79 tests**

```bash
# Run all planner tests
pytest tests/planner/ -v -o addopts=""

# Breakdown:
# - test_workflow_builder.py:    15 tests
# - test_field_mapping.py:        14 tests
# - test_registry.py:             23 tests
# - test_llm_planner.py:          16 tests
# - test_workflow_format.py:      11 tests
```

**Test Coverage**:
- Workflow building and serialization
- Field mapping (explicit, LLM, exact match)
- Agent registry integration
- LLM planner conversation flow
- Input extraction with full context
- Workflow format validation (io_map, edges, structure)
- Error handling and edge cases

## Files Changed

**Added**:
- akd/planner/format_builder.py
- akd/planner/workflow_builder.py
- akd/planner/field_mapping_registry.py
- akd/planner/field_mapping_generator.py
- akd/planner/structures.py
- akd/planner/README.md
- akd/mapping/field_mappings.json
- scripts/minimal_planner.py
- tests/planner/test_workflow_builder.py
- tests/planner/test_field_mapping.py
- tests/planner/test_workflow_format.py

**Modified**:
- akd/planner/__init__.py (added exports)
- akd/agents/search/_base.py (added report field)
- akd/agents/search/deep_search.py (populate report)
- akd/mapping/agent_registry.json (updated schema)

## Usage

### Quick Start - LLM Planner

```python
from akd.planner.llm_planner import quick_plan

# Quick planning (non-interactive)
session = await quick_plan("Find recent papers on AlphaFold protein structure prediction")
response = await session.start()
workflow = await session.generate_workflow()

# Save to file
workflow.save_to_file("my_workflow.json")
```

### Interactive Planning (NEW)

```python
from akd.planner.llm_planner import create_planner

# Create planner
planner = await create_planner()

# Start interactive session
session = await planner.plan_workflow("Find papers on AlphaFold and identify research gaps")
response = await session.start()

# Conversation loop
while not response.ready_to_generate:
    print(f"Planner: {response.message}")
    user_input = input("Your response: ")
    response = await session.respond(user_input)

# Generate workflow
workflow = await session.generate_workflow()
workflow.save_to_file("alphafold_workflow.json")
```

### Backend Integration Example

`scripts/minimal_planner.py` demonstrates backend/frontend integration:

```bash
python3 scripts/minimal_planner.py "Find papers on AlphaFold and identify research gaps"
```

### Direct WorkflowBuilder Usage

```python
from akd.planner import (
    get_agent_registry,
    WorkflowBuilder,
    WorkflowPlan,
    AgentSuggestion,
)

# Initialize (uses registry from PR #236)
registry = get_agent_registry()
builder = WorkflowBuilder(registry)

# Create plan
plan = WorkflowPlan(
    workflow_description="Find papers then analyze gaps",
    research_goal="Identify research gaps in AlphaFold",
    suggested_agents=[
        AgentSuggestion(
            agent_id="deep_search",
            agent_name="Deep Literature Search",
            reason="Search for papers",
            confidence=0.95,
        ),
        AgentSuggestion(
            agent_id="gap_analysis",
            agent_name="Gap Analysis",
            reason="Analyze gaps",
            confidence=0.90,
        )
    ],
    estimated_complexity="medium"
)

# Build workflow with auto-generated io_map
workflow = builder.build(plan, {
    "deep_search": {"query": "AlphaFold"}
})

# Result: gap_analysis.search_results auto-mapped to $.deep_search.outputs.results
print(workflow.to_json())
```

## Interactive Demo

Try the interactive CLI:

```bash
# Interactive planning session
python scripts/demo_planner.py interactive

# Quick planning
python scripts/demo_planner.py quick "Find papers on protein folding"

# Show available agents
python scripts/demo_planner.py agents

# Sample workflow format
python scripts/demo_planner.py sample
```
## Documentation

See comprehensive documentation at `akd/planner/README.md`.

## Enabled Workflows

```
deep_search → gap_analysis
code_search → gap_analysis
deep_search → code_search → gap_analysis
```

muthukumaranR committed:
Add field mappings configuration for agent interactions

muthukumaranR committed:
Add synthesized research report field to LitSearchAgentOutputSchema and update related agent logic

- Introduced a new field `report` in `LitSearchAgentOutputSchema` to store synthesized research reports.
- Updated `DeepLitSearchAgent` to pass the `research_report` to the new `report` field.
- Modified `agent_registry.json` to include the new `report` field in the agent configuration.
- Refactored test cases in `deep_search_test.py` to reflect changes in the output schema and ensure proper validation of the report field.

muthukumaranR committed:
Add workflow format builder module for AKD planner

- define classes for building workflow definitions in the AKD framework.
- Implement `WorkflowField`, `WorkflowNodeIO`, `WorkflowNode`, `WorkflowEdge`, and `WorkflowFormat` classes for structured workflow representation.
- Added methods for custom serialization, JSON export, and file saving for workflow definitions.
- Included functions to create and load workflows from dictionaries and JSON files, respectively.

muthukumaranR committed:
Add workflow builder for AKD planner

- Introduced `WorkflowBuilder` class to construct `WorkflowFormat` from `WorkflowPlan` with runtime data flow.
- Implemented a three-tier field mapping strategy: explicit mappings, exact name matches, and LLM-generated mappings.
- Added methods for building nodes and edges, validating agent existence, and identifying unmapped fields for LLM generation.
- Integrated logging for better traceability of mapping processes and warnings for missing agents or fields.

NISH1001 commented on akd/agents/search/_base.py:
```diff
@@ -89,6 +89,10 @@ class LitSearchAgentInputSchema(SearchAgentInputSchema):
 class LitSearchAgentOutputSchema(SearchAgentOutputSchema):
     """Base output schema for literature search agents."""
 
+    report: str = Field(
+        default="",
+        description="Synthesized research report from the literature search",
+    )
```
Nice.

muthukumaranR committed:
Refactor deep search tests to reflect changed output schema for report

NISH1001 commented on akd/planner/format_builder.py:
```diff
@@ -0,0 +1,100 @@
+class WorkflowNodeIO(BaseModel):
+    """Input or output specification for a workflow node."""
+
+    fields: List[Union[Dict[str, Any], WorkflowField]] = Field(default_factory=list)
```
Should we use modern Python 3.12 typing style? Union -> `|`, `List->list`, etc?

NISH1001 commented on akd/planner/format_builder.py:
```diff
@@ -0,0 +1,100 @@
+    type: str = Field(..., description="Node type corresponding to agent type")
+    input: WorkflowNodeIO = Field(default_factory=WorkflowNodeIO)
+    output: Optional[WorkflowNodeIO] = Field(default=None)
+    io_map: Optional[Dict[str, str]] = Field(
```
Nice.

NISH1001 commented on akd/planner/format_builder.py:
```diff
@@ -0,0 +1,100 @@
+        """Export workflow to JSON string."""
+        # Use exclude_none=True by default for cleaner output (omit null fields)
+        kwargs.setdefault("exclude_none", True)
+        return json.dumps(self.model_dump(**kwargs), indent=2)
```
Doesn't pydantic v2+ has `model_dump(mode="json")` mode as well? Will that work? and to_json will be just a proxy method?

NISH1001 commented on akd/planner/format_builder.py:
```diff
@@ -0,0 +1,100 @@
+            f.write(self.to_json(**kwargs))
+
+
+def create_workflow_from_dict(data: Dict[str, Any]) -> WorkflowFormat:
```
Maybe this should be `WorkflowFormat.from_dict(...)` classmethod no? Tightly couples the method.

```python
class WorkflowFromat(BaseModel):
  ...
  ...
  @classmethod
  def from_dict(cls, ...) -> WorkflowFormat:
...
```

```python
# usage

workflow_format = WorkflowFormat.from_dict(...)
```

NISH1001 commented on akd/planner/format_builder.py:
```diff
@@ -0,0 +1,100 @@
+    """Load a workflow from a JSON file."""
+    with open(file_path, "r") as f:
+        data = json.load(f)
+    return create_workflow_from_dict(data)
```
Let's put this as another classmethod to WorkflowFormat

```python
class ClassWorkflowFormat...
  ...
  ...

  @clasmethod
  def from_file(cls, file_path:str)...
...



# usage

workflow_format = WorkflowFormat.from_file(...)
```

NISH1001 commented on akd/planner/workflow_builder.py:
```diff
@@ -0,0 +1,211 @@
+
+    def __init__(
+        self,
+        registry: AgentRegistry,
```
Should registry be optional and we can do `self.registry = registry or AgentRegistry() ? Becuase there's only 1 registry, so default should be that?

NISH1001 commented on akd/planner/workflow_builder.py:
```diff
@@ -0,0 +1,211 @@
+                                        f"No mapping found for {agent_id}.{field.name} from "
+                                        f"{prev_agent_id}. Field will be missing unless LLM mapping "
+                                        f"is generated."
+                                    )
```
Should we move this to an internal method? It will help clean the build(...) method for readability?

NISH1001 commented on akd/planner/workflow_builder.py:
```diff
@@ -0,0 +1,211 @@
+
+        return WorkflowFormat(
+            workflow_type="AKDResearchWorkflow",
+            version="1.0.0",
```
maybe we can get this from somewhere? Like `WorkflowBuilder.__version__`? Or even project config?

NISH1001 replied:
This is now handled by `akd.planner.WORKFLOW_TYPE` and `akd.planner.WORKFLOW_VERSION` global consts.

NISH1001 commented on akd/planner/workflow_builder.py:
```diff
@@ -0,0 +1,211 @@
+            output=nodes[-1].output if nodes else None
+        )
+
+    def check_agents_exist(self, plan: "WorkflowPlan") -> List[str]:
```
I think this name is misleading? Should we rename it something like `check_missing_agents(...)`?

NISH1001 reviewed (changes requested):
Some comments: Some are nitpick. Some are improvements.

NISH1001 commented:
REsolving some of these comments myself.

NISH1001 committed:
Refactor type hint for workflownode to use ne wpython style

- ALos use pydantic-v2 style model config `model_config` intead of
  `class Config`

NISH1001 committed:
Bugfix tests for deep search agent

- Bugfix `report` field access

NISH1001 committed:
Refactor WorkflowFormat to have class methods

NISH1001 commented on akd/planner/workflow_builder.py:
```diff
@@ -0,0 +1,211 @@
+
+from .format_builder import WorkflowFormat, WorkflowNode, WorkflowEdge, WorkflowNodeIO
+from .registry import AgentRegistry
+from .field_mapping_registry import FieldMappingRegistry
```
This module doens't exist apparently.

NISH1001 commented on akd/planner/workflow_builder.py:
```diff
@@ -0,0 +1,211 @@
+from .field_mapping_registry import FieldMappingRegistry
+
+if TYPE_CHECKING:
+    from .llm_planner import WorkflowPlan
```
This module doens't exist apparently.

muthukumaranR committed:
Add field mapping registry and generator modules

- FieldMappingRegistry: Three-tier mapping resolution (explicit, exact match, LLM)
- FieldMappingGenerator: LLM-based semantic field mapper
- Required by WorkflowBuilder for intelligent field mapping between agents

muthukumaranR committed:
Add pytest tests for field mapping and workflow builder

- test_field_mapping.py: 14 tests for FieldMappingRegistry and FieldMappingGenerator
- test_workflow_builder.py: 15 tests for WorkflowBuilder
- All tests passing (29 total)
- Covers explicit mappings, LLM mappings, confidence thresholds, workflow generation

muthukumaranR committed:
Refactor field mapping generator by removing unused classes and updating imports (moved to structures)

muthukumaranR committed:
Update WorkflowBuilder to use AbstractWorkflowPlan type for plan parameter - (moving the Workflow Builder for the PR accompanying llm_planner

muthukumaranR committed:
Add data structures for AKD planner workflow management

muthukumaranR committed:
Update import statements in test_workflow_builder.py to reflect new module structure for WorkflowPlan and AgentSuggestion

muthukumaranR committed:
Update import statement in test_field_mapping.py to align with new module structure for WorkflowPlan and AgentSuggestion

NISH1001 committed:
Bugfix tests for planner registry

NISH1001 committed:
Make agent registry optoinal in workflowbuilder

- Also add debug flag

NISH1001 committed:
Refactor build() method in workflowbuilder to separate out io mapping
method

NISH1001 committed:
REname agent check method to missing agent check in workflowbuilder

NISH1001 committed:
Add WORKFLOW_TYPE and WORKFLOW_VERSION global constants to planner

muthukumaranR committed:
Add imports for FieldMappingGenerator, FieldMappingRegistry, WorkflowBuilder, AgentSuggestion, and WorkflowPlan in planner module

muthukumaranR committed:
Add abstract planner interfaces to make PR #2 self-contained

- Add AbstractWorkflowPlanner, AbstractInteractivePlanner, AbstractPlannerSession
- Add PlannerConfig to structures.py
- Update __init__.py exports with all abstract interfaces
- Add tests for abstract planner interfaces (9 tests passing)
- Makes PR #2 complete without depending on PR #3 implementation
- PR #3 (LLM Planner) will implement these abstract interfaces

Total tests: 61 passing (15 workflow_builder + 14 field_mapping + 23 registry + 9 base_planner)

muthukumaranR committed:
Improve type safety in planner module - remove Any types

- Remove AbstractWorkflowPlan (unused, had List[Any])
- Add ComplexityLevel type alias (Literal['low', 'medium', 'high'])
- Add PlannerContext TypedDict for planner context parameter
- Add UnmappedFieldInfo TypedDict for unmapped field data
- Replace Dict[str, Any] with specific types throughout
- Replace Any with AgentEntry in workflow_builder
- Replace Any with FieldValue union type in format_builder
- Change all List[] to list[] for consistency
- Add bounds (ge, le) to confidence fields

Benefits:
- Better type checking and IDE support
- Clearer API contracts
- No loss of flexibility where needed
- All 38 tests passing

muthukumaranR committed:
Simplify type system: replace JsonValue with str, use akd.serializers

- Remove JsonValue type alias (Union of JSON-serializable types)
- Change PlannerContext.previous_results to dict[str, str] (JSON strings)
- Update docstring to reference akd.serializers for serialization
- Remove JsonValue from exports in __init__.py
- Fix bug in llm_planner.py: check_agents_exist -> check_missing_agents
- Replace generic object type with specific types in format_builder.py

All 77 planner tests passing.

muthukumaranR committed:
Remove ComplexityLevel type alias - inline Literal constraint

- ComplexityLevel was only used in one field (WorkflowPlan.estimated_complexity)
- Inline Literal["low", "medium", "high"] directly in field definition
- Remove from exports in __init__.py to simplify public API
- Reduce unnecessary type definitions per user feedback

All 77 tests passing.

muthukumaranR committed:
Remove estimated_complexity field - unused decorative metadata

- Remove estimated_complexity from WorkflowPlan (only used for display)
- Remove from LLM system prompt
- Remove from conversation summary display
- Remove Literal import (no longer needed)
- Update all test fixtures to remove the field

No business logic depended on this field - it was purely decorative.

All 77 tests passing.

muthukumaranR committed:
Refactor LLM workflow planner to enhance type safety and modularity

- Replace InstructorBaseAgent with LiteLLMInstructorBaseAgent for improved performance.
- Update input and output schemas to use list and dict types instead of List and Dict for consistency.
- Introduce PlannerConfig for planner-specific configurations.
- Optimize system prompt generation using templates for better maintainability.
- Enhance input extraction logic to utilize full conversation context and workflow plan.

muthukumaranR committed:
Update AKD Planner module documentation and exports

muthukumaranR committed:
Refactor LLM Planner to bypass mapped fields in output and improve context management

- Make all fields optional to allow LLM to omit auto-mapped fields, simplifying input handling.
- Introduce logic to check for available field mappings from previous agents, improving context awareness.
- Update context section to include information about auto-mapped fields, ensuring clarity in output expectations.
- Filter out None values from the result to clean up the output, enhancing data integrity.

muthukumaranR committed:
Enhance WorkflowBuilder input validation and logging

- Introduce validation for source fields in WorkflowBuilder to ensure mappings reference existing output fields from previous agents.

muthukumaranR committed:
Refactor field mapping types for consistency

- Replace Dict with dict for type annotations in field_mapping_registry.py to align with modern Python typing conventions.
- Update type hints in LLMMappingEntry and FieldMappingRegistry classes for improved type safety and clarity.
- Ensure all mappings utilize consistent type definitions, enhancing maintainability and readability.

muthukumaranR committed:
Refactor type annotations in field_mapping_generator.py for consistency

- Replace List with list and Dict with dict in type hints to align with modern Python typing conventions.
- Enhance type safety and clarity in method signatures, ensuring consistent usage of type definitions throughout the file.
- Update comments to reflect changes in type annotations, improving maintainability and readability.

muthukumaranR committed:
Remove abstract base planner classes and associated tests

- Delete the abstract base classes for workflow planners from base_planner.py, which defined interfaces for future implementations.
- Remove the corresponding test file test_base_planner.py that validated the abstract interfaces and configurations.
- This cleanup prepares the codebase for upcoming concrete implementations of workflow planners.

muthukumaranR committed:
Enhance WorkflowPlan validation and configuration

- Introduce a field validator for 'workflow_steps' in WorkflowPlan to ensure steps reference actual agents, logging any plain language descriptions at DEBUG level
- Update PlannerConfig to include a new confidence threshold for auto-approving field mappings
- adjust the temperature setting for deterministic planning.
- Improve type safety and maintainability by adhering to modern Python typing conventions.

muthukumaranR committed:
Add demo planner script for interactive and automated workflow planning

muthukumaranR committed:
Add prompt templates for workflow planner module

muthukumaranR committed:
Add README.md for AKD Planner Module

muthukumaranR committed:
Refactor imports in test_llm_planner.py for improved organization

muthukumaranR committed:
fix ruff formatting issues

muthukumaranR committed:
fix ruff formatting

muthukumaranR committed:
ruff formatting v2

muthukumaranR committed:
Update README.md to enhance documentation for the Workflow Planning System

muthukumaranR committed:
Update workflow generation strategy in planner_prompts.py

muthukumaranR committed:
Add  validation methods to WorkflowFormat class

muthukumaranR committed:
Implement field validators for question_type and phase normalization, auto-set ready_to_generate based on workflow_plan existence, and rename plan_workflow to init_planner_session for clarity in LLMWorkflowPlanner class. Enhance workflow validation logging in InteractivePlannerSession.

muthukumaranR committed:
Refactor import statements and enhance code formatting in demo_planner.py for improved readability. Update method calls to use init_planner_session for clarity and ensure consistent string formatting throughout the file.

muthukumaranR committed:
Add minimal interactive planner script for backend/frontend integration.

muthukumaranR committed:
Add unit tests for WorkflowFormat validation methods in test_workflow_format.py to ensure proper detection of invalid node references, JSONPath formats, orphaned nodes, and edge cases. Enhance coverage for both strict and non-strict validation scenarios.

NISH1001 committed:
Add conversation_history parameter to planner interactive session

- `conversation_history: list[dict[str, str]]` is added to
  INteractivePlannerSession
- So, that externally we can inject the history for resuming

NISH1001 commented on akd/planner/field_mapping_registry.py:
```diff
@@ -0,0 +1,340 @@
+
+            # Convert to dict for JSON serialization
+            data = {
+                "version": "1.0.0",
```
Need this to be switched to `WORKFLOW_FORMAT_VERSION` for flexibility in newer versioning maybe?

NISH1001 commented on akd/planner/field_mapping_registry.py:
```diff
@@ -0,0 +1,340 @@
+            explicit_path: Path to explicit mappings JSON file
+            llm_path: Path to LLM-generated mappings JSON file
+        """
+        self.explicit_path = explicit_path or "akd/mapping/field_mappings.json"
```
This should be based off akd root. 

```python
from akd.utils import get_akd_root

....explicit_path or get_akd_root() / "mapping/field_mappings.json"
```

NISH1001 commented on akd/planner/field_mapping_registry.py:
```diff
@@ -0,0 +1,340 @@
+            llm_path: Path to LLM-generated mappings JSON file
+        """
+        self.explicit_path = explicit_path or "akd/mapping/field_mappings.json"
+        self.llm_path = llm_path or "akd/mapping/llm_generated_mappings.json"
```
Same as above:

This should be based off akd root.

```
from akd.utils import get_akd_root

....path or get_akd_root() / "mapping/............".....
```

NISH1001 committed:
Validate jsonpath in workflow builder for planner

NISH1001 committed:
Change hardcoded path in field_mapping_registry to use
akd.utils.get_akd_root

NISH1001 committed:
Inject conversation_history to init_planner_session method

NISH1001 committed:
Use default factory for llm-generated field mapping timestamp

- `LLMMappingEntry.generated_at` now uses default factory
